### PR TITLE
[openvino] Support PReLU and Activation ops

### LINKED ISSUE #1002: PReLU op support for Face-Reidentification and ESRGAN
Face-Reidentification model has PReLU op in the graph.

PReLU op:

Screenshot from 2019-09-29 15-04-41

Face-Reidentification model need ops table:

Screenshot from 2019-09-29 15-04-24

As a PReLU op, NNAPI has **ANEURALNETWORKS_PRELU** op, and ONNX has **PRelu** op.

<table>
    <tr>
        <th>Type</th>
        <th>Detail</th>
        <th>NNAPI</th>
	<th>ONNX</th>
    </tr>
    <tr>
        <td rowspan="4">PReLU</td>
	<td>Op</td>
        <td><a href="https://developer.android.com/ndk/reference/group/neural-networks.html#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a89be86ff36e83254e9b3d0954e7e0729">ANEURALNETWORKS_PRELU</a></td>
	<td><a href="https://github.com/onnx/onnx/blob/master/docs/Operators.md#PRelu">    PRelu</a></td>
    </tr>
    <tr>
        <td>Input</td>
        <td><p>input[0]: <br>Supported tensor rank: from 1. <br>A tensor, specifying the input.</p></td>
	<td><p>inputs[“X”]: An input tensor.</p></td>
    </tr>
    <tr>
        <td>Slope</td>
        <td><p>input[1]: <br>A tensor of the same OperandCode, and compatible dimensions as input0, specifying the alpha.</p></td>
	<td><p>inputs["slope "]:<br>Slope tensor. The shape of slope can be smaller then first input X; if so, its shape must be unidirectional broadcastable to X.</p></td>
    </tr>
	<tr>
        <td>Output</td>
        <td><p>output[0]: <br>A tensor of the same OperandCode as input0. For a ANEURALNETWORKS_TENSOR_QUANT8_ASYMM tensor, the scale and zeroPoint can be diffent from the input0 scale and zeroPoint.</p></td>
		<td><p>outputs["Y "]:<br>Output tensor (same size as inputs[“X”]).</p></td>
    </tr>
</table>​

### LINKED ISSUE #1006: Sigmoid (logistic) op support for Face-Reidentification and DeepSpeech
Face-Reidentification model has Sigmoid op in the graph.

Activation(Sigmoid) op:

Screenshot from 2019-09-29 15-06-18

Face-Reidentification model need ops table:

Screenshot from 2019-09-29 15-04-24

As a  Activation(Sigmoid) op, NNAPI has  **ANEURALNETWORKS_LOGISTIC** op, and ONNX has  **Sigmoid** op.

<table>
    <tr>
        <th>Type</th>
        <th>Detail</th>
        <th>NNAPI</th>
	<th>ONNX</th>
    </tr>
    <tr>
        <td rowspan="3">Sigmoid</td>
	<td>Op</td>
        <td><a href="https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a82a340eb540933f638db420369650483">ANEURALNETWORKS_LOGISTIC</a></td>
	<td><a href="https://github.com/onnx/onnx/blob/master/docs/Operators.md#Sigmoid">Sigmoid</a></td>
    </tr>
    <tr>
        <td>Input</td>
        <td><p>input[0]: <br>Supported tensor rank: up to 4. <br>A tensor, specifying the input.</p></td>
	<td><p>inputs[“X”]: Input tensor.</p></td>
    </tr>
    <tr>
        <td>Output</td>
        <td><p>output[0]: <br>A tensor of the same OperandCode as input0. For a ANEURALNETWORKS_TENSOR_QUANT8_ASYMM tensor, the scale must be 1.f / 256 and the zeroPoint must be 0.</p></td>
		<td><p>outputs["Y "]:<br>Output tensor.</p></td>
    </tr>
</table>​

### PR DESCRIPTION
**Pull request by cuiyanx**
1. Support **Normalize** and **PReLU** and **Activation** ops.
2. Implement Int32 data type.
3. Recode **Permute** ops.
4. Fix none-value and dims-value bugs for **FullyConnected** op.

fixes #1002 and fixes #1006

Signed-off-by: cuiyan <yanx.cui@intel.com>

### CODE REVIEW

**Comment by cuiyanx** [comment]
@ibelem @BruceDai please review

**Comment by cuiyanx** [comment]
No need to be supported with Normalize op.
fixes #1001

@ibelem @BruceDai please review again.

File: examples/util/openvino/OpenVINOModelImporter.js [hunk 1]
```diff
@@ -681,6 +737,90 @@ class OpenVINOModelImporter {
...
         } break;
+        case 'Normalize': {
+          const input = node.inputs[0];
+          const inDims = input.shape();
+          const inputId = this._getTensorId(input);
+          console.log(`  input shape: [${input.shape()}]`);
+
+          const acrossSpatial = node.getInt('across_spatial');
+          const channelShared = node.getInt('channel_shared');
+          const eps = node.getFloat('eps');
```
Comment (huningxin) [inline examples/util/openvino/OpenVINOModelImporter.js]: do we support `eps`?

File: examples/util/openvino/OpenVINOModelImporter.js [hunk 2]
```diff
@@ -681,6 +737,90 @@ class OpenVINOModelImporter {
...
+
+          if (acrossSpatial !== 0 || channelShared !== 0) {
+            throw new Error(`Normalize not support across_spatial ${across_spatial} channel_shared ${channel_shared}.`);
+          }
+
+          inputs.push(inputId);
+
+          const output = node.outputs[0];
+
+          if (i === graph.nodes.length - 1) {
```
Comment (huningxin) [inline examples/util/openvino/OpenVINOModelImporter.js]: could you please elaborate why can skip the last `Normalize` layer?

File: examples/util/openvino/OpenVINOModelImporter.js [hunk 3]
```diff
@@ -681,6 +737,90 @@ class OpenVINOModelImporter {
...
+              type: this._getTypeCode(output.dataType()), dimensions: outDims
+            };
+            const outputId = this._addNamedOperand(output.graphId(), outputType);
+            outputs.push(outputId);
+            console.log(`  output shape: [${outDims}]`);
+
+            opCode = this._nn.L2_NORMALIZATION;
+          }
+        } break;
+        case 'PReLU': {
```
Comment (huningxin) [inline examples/util/openvino/OpenVINOModelImporter.js]: should we handle `channel_shared` parameter of [PReLU](https://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_convert_model_IRLayersCatalogSpec.html#PReLU)?

File: examples/util/openvino/OpenVINOModelImporter.js [hunk 4]
```diff
@@ -681,6 +737,90 @@ class OpenVINOModelImporter {
...
+          opCode = this._nn.PRELU;
+        } break;
+        case 'Activation': {
+          // Add inputs
+          const in1 = node.inputs[0];
+          inputs.push(this._getTensorId(in1));
+          console.log(`  inputs shape: ` +
+              `[${node.inputs.map((input) => input.shape()).join('], [')}]`);
+
+          const operation = node.getString('type');
```
Comment (huningxin) [inline examples/util/openvino/OpenVINOModelImporter.js]: better to name the variable as `type`.

File: examples/util/openvino/OpenVINOModelImporter.js [hunk 5]
```diff
@@ -681,6 +737,90 @@ class OpenVINOModelImporter {
...
+          console.log(`  operation: ${operation}`);
+          switch (operation) {
+            case 'sigmoid':
+              opCode = this._nn.LOGISTIC;
+              break;
+            default:
+              throw new Error(`Operation ${operation} is not supported`);
```
Comment (huningxin) [inline examples/util/openvino/OpenVINOModelImporter.js]: Better to show error message as "The type ${type} of Activation is not supported"

File: examples/util/openvino/OpenVINOModelImporter.js (hunk 1 above)
Comment (cuiyanx) [reply examples/util/openvino/OpenVINOModelImporter.js]: No, the NNAPI not support, so we do not.

File: examples/util/openvino/OpenVINOModelImporter.js (hunk 2 above)
Comment (cuiyanx) [reply examples/util/openvino/OpenVINOModelImporter.js]: Only in this case, when the Normalize operation appears as the last operation， we can skip the Normalize operation. and I realized this operation as function in [examples/face_recognition/utils.js](https://github.com/intel/webml-polyfill/pull/984/files?file-filters%5B%5D=.html&file-filters%5B%5D=.jpg&file-filters%5B%5D=.js#diff-8095bd061b3d471362cc188b106f9bec) of PR #984.

File: examples/util/openvino/OpenVINOModelImporter.js (hunk 3 above)
Comment (cuiyanx) [reply examples/util/openvino/OpenVINOModelImporter.js]: No, there are only tow inputs as input and slop in [NNAPI](https://developer.android.com/ndk/reference/group/neural-networks.html#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a82e7e7c709ad117a0178203217be30cf), and there is auto-handle channel_shared parameter in NNAPI,  so we no need this parameter.

File: examples/util/openvino/OpenVINOModelImporter.js (hunk 1 above)
Comment (huningxin) [reply examples/util/openvino/OpenVINOModelImporter.js]: Then I suggest to remove this line and add a comment that says we don't support `eps` now.

File: examples/util/openvino/OpenVINOModelImporter.js (hunk 2 above)
Comment (huningxin) [reply examples/util/openvino/OpenVINOModelImporter.js]: For this case, I would suggest to extract the `Normalization` layer from your model instead adding `Normalization` support.

File: examples/util/openvino/OpenVINOModelImporter.js (hunk 3 above)
Comment (huningxin) [reply examples/util/openvino/OpenVINOModelImporter.js]: Then, please add comment to say we don't support `channel_shared` parameter of OpenVINO's PReLU.

File: examples/util/openvino/OpenVINOModelImporter.js (hunk 1 above)
Comment (cuiyanx) [reply examples/util/openvino/OpenVINOModelImporter.js]: I delete Normalization supportting.

File: examples/util/openvino/OpenVINOModelImporter.js (hunk 2 above)
Comment (cuiyanx) [reply examples/util/openvino/OpenVINOModelImporter.js]: I delete Normalization supportting.

**Comment by cuiyanx** [comment]
@huningxin I delete `Normalize` op supportting and add some comment, please review

File: examples/util/openvino/OpenVINOModelImporter.js [hunk 6]
```diff
@@ -620,15 +626,65 @@ class OpenVINOModelImporter {
...
+
+              const outDims = output.shape();
+              const outputType = {
+                type: this._getTypeCode(output.dataType()), dimensions: outDims
+              };
+              const outputId = this._addNamedOperand(outputName, outputType);
+              outputs.push(outputId);
+              console.log(`  output shape: [${outDims}]`);
+
+              this._addOperation(this._nn.TRANSPOSE, inputs, outputs);
```
Comment (huningxin) [inline examples/util/openvino/OpenVINOModelImporter.js]: Why do we need 3 `TRANSPOSE` ops? Can we map the axis between NHWC and NCHW and use only one transpose?

Comment (cuiyanx) [reply examples/util/openvino/OpenVINOModelImporter.js]: @huningxin Yes, we can. I transpose `input` data, but you want to transpose `order` data and use only one op. So I will update this op.

**Comment by cuiyanx** [comment]
@huningxin I rebase and recode `Permute` op, please review again.

**Comment by cuiyanx** [comment]
@huningxin would you review the newest commit? thanks.

**Comment by huningxin** [comment]
Looks good. And thanks much for your efforts to continuously improve the PR in all the review process.

**Pull request merged**